sdks/python/apache_beam/transforms/sql.py

```diff
@@ -67,10 +66,9 @@ class SqlTransform(ExternalTransform):
   """
   URN = 'beam:external:java:sql:v1'
 
-  def __init__(self, query, dialect=None):
+  def __init__(self, query, dialect=None, expansion_service=None):
     super(SqlTransform, self).__init__(
         self.URN,
         NamedTupleBasedPayloadBuilder(
             SqlTransformSchema(query=query, dialect=dialect)),
-        BeamJarExpansionService(
-            ':sdks:java:extensions:sql:expansion-service:shadowJar'))
+        expansion_service=expansion_service)
```

Comment:
We definitely want to accept an expansion service as an argument, but the default should remain `BeamJarExpansionService(':sdks:java:extensions:sql:expansion-service:shadowJar')`.

Reply:
Note that you can use `JavaJarServer.beam_services()` (automatically populated by the `--beam_services` flag) to specify an alternative jar or live expansion service for any beam target.

Reply:
Thanks for the pointer. The workaround via overriding the Gradle target may be acceptable for developers but exposing it via an argument to users does not seem right. Users shouldn't have to know Gradle to use Beam Python. 

Also, the corresponding value can either be a URL or a `host:port` string which is confusing to users and probably will break for Ipv6 addresses.

Reply:
Yes, the `beam_services` flag is primarily for developers/testing. The default should just work after running "pip install apache_beam" which is what `BeamJarExpansionService` aims to provide. Externally-started job servers running universal expansion services at a fixed port was useful for bootstrapping, but is a bit of a special configuration. Exactly how to best configure this (which, in large part, boils down to being able to name expansion (or other) services), is a good open question.

Reply:
If the goal is to make things easier/flexible for developers, then convenience built into Gradle tasks may be a good option. Users would probably find it rather problematic if by default there is an attempt to pull down artifacts behind the scenes and execute them. That is probably also going to fail within isolated production environments.

Reply:
For a good out-of-the-box experience it may be acceptable to bundle the expansion service with the job server, which is already the case for the Flink Runner. That way we don't have to start a separate service. I was thinking to move the dependencies for expansion (SQL, Kafka, etc.) to the base job-service project (https://issues.apache.org/jira/browse/BEAM-10386), but decided to postpone it for now because it causes a larger refactoring. IMHO we should get rid of the Runner-specific job service projects entirely, each Runner's artifact should carry all the necessary dependencies through the central job-service project. 

The current approach here is very modular but downloading jars is unexpected because most users would expect everything to be self-contained. The downside also is that it won't easily work for other languages. In the future we may want to either (a) include a self-contained expansion server in the final artifact (b) publish container images for all expansion servers.

I'll keep the current default because this warrants a mailing list discussion.

Reply:
+1, a wider discussion about expansion service discovery and bundling would be very good to have on the mailing list.